src/app/clusters/closure-dimension-server/closure-dimension-cluster-logic.cpp

```diff
@@ -19,93 +19,302 @@
     {
+        //  If the speed member is present in the incoming CurrentState, we need to check if the Speed feature is
+        //  supported by the closure. If the Speed feature is not supported, return an error.
         VerifyOrReturnError(mConformance.HasFeature(Feature::kSpeed), CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE);
-        VerifyOrReturnError(EnsureKnownEnumValue(currentState.speed.Value()) != Globals::ThreeLevelAutoEnum::kUnknownEnumValue,
+
+        const Globals::ThreeLevelAutoEnum & speed = incomingCurrentState.Value().speed.Value();
+
+        VerifyOrReturnError(EnsureKnownEnumValue(speed) != Globals::ThreeLevelAutoEnum::kUnknownEnumValue,
                             CHIP_ERROR_INVALID_ARGUMENT);
+
+        // We need to update the cluster position value if
+        // - cluster does not have value; the value will necessarily be updated
+        // - cluster and incoming values are different
+        requireSpeedUpdate = !validateClusterCurrentStateSpeed || clusterCurrentState.Value().speed.Value() != speed;
     }
 
-    // TODO: currentState.Position value SHALL follow the scaling from "Resolution Attribute".
+    /*
+        Update the cluster state and mark it as dirty if necessary
+    */
 
-    if (currentState != mState.currentState)
+    // Cluster CurrentState going from non-null to null
+    if (validateClusterCurrentStateMembers && !validateIncomingCurrentStateMembers)
+    {
+        clusterCurrentState.SetNull();
+        mMatterContext.MarkDirty(Attributes::CurrentState::Id);
+    }
+    else if (requirePositioningUpdate || requireLatchUpdate || requireSpeedUpdate)
     {
-        mState.currentState = currentState;
+        clusterCurrentState.SetNonNull(incomingCurrentState.Value());
         mMatterContext.MarkDirty(Attributes::CurrentState::Id);
     }
 
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR ClusterLogic::SetTarget(const GenericTargetStruct & target)
+CHIP_ERROR ClusterLogic::SetTarget(const DataModel::Nullable<GenericTargetStruct> & incomingTarget)
 {
+    assertChipStackLockedByCurrentThread();
+
     VerifyOrReturnError(mInitialized, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mState.target != incomingTarget, CHIP_NO_ERROR);
+
+    DataModel::Nullable<GenericTargetStruct> & clusterTarget = mState.target;
 
-    if (target.position.HasValue())
+    bool validateClusterTargetMembers  = false;
+    bool validateClusterTargetPosition = false;
+    bool validateClusterTargetLatch    = false;
+    bool validateClusterTargetSpeed    = false;
+
+    bool validateIncomingTargetMembers  = false;
+    bool validateIncomingTargetPosition = false;
+    bool validateIncomingTargetLatch    = false;
+    bool validateIncomingTargetSpeed    = false;
+
+    bool requirePositioningUpdate = false;
+    bool requireLatchUpdate       = false;
+    bool requireSpeedUpdate       = false;
+
+    /*
+        Determine checks that need to be executed based on current and provided values
```

Comment:
Again, what is the actual desired behavior here?  This seems way over-complicated.